browser/brave_news/brave_news_tab_helper.cc

```diff
@@ -89,48 +102,89 @@ bool BraveNewsTabHelper::IsSubscribed(const FeedDetails& feed_details) {
 }
 
 bool BraveNewsTabHelper::IsSubscribed() {
-  for (const auto& feed : GetAvailableFeeds()) {
-    if (IsSubscribed(feed))
+  for (const auto& feed_url : GetAvailableFeedUrls()) {
+    if (IsSubscribed(feed_url)) {
       return true;
+    }
   }
   return false;
 }
 
-void BraveNewsTabHelper::ToggleSubscription(const FeedDetails& feed_details) {
-  bool subscribed = IsSubscribed(feed_details);
-  auto* publisher = controller_->publisher_controller()->GetPublisherForFeed(
-      feed_details.feed_url);
+std::string BraveNewsTabHelper::GetTitleForFeedUrl(const GURL& feed_url) {
+  auto* default_publisher =
+      controller_->publisher_controller()->GetPublisherForFeed(feed_url);
+  if (default_publisher) {
+    return default_publisher->publisher_name;
+  }
+
+  auto it = base::ranges::find_if(
+      rss_page_feeds_,
+      [feed_url](const auto& details) { return details.feed_url == feed_url; });
+  if (it == rss_page_feeds_.end()) {
+    return "";
+  }
+
+  if (it->title.empty()) {
+    const auto url = web_contents()->GetLastCommittedURL();
+    controller_->FindFeeds(
```

Comment:
I see that `DirectFeedController::FindFeeds` has a guard for preventing multiple calls for the same feed_url resulting in multiple network requests. The callbacks are queued and all called at the same time once the first (and only) request for that feed_url is complete.
However, it seems like we could at least prevent this getting queued, since we don't need our callback to get run multiple times since it's doing the same thing each time. We just need `OnFoundFeeds` to get called once. Should we protect against calling FindFeeds once we've called it for this `feed_url`?

Reply:
Sweet, I'm adding a `requested_feed` boolean to the `FeedDetails` struct.